compiler/rustc_const_eval/src/transform/validate.rs

```diff
@@ -301,9 +302,16 @@ impl<'a, 'tcx> Visitor<'tcx> for TypeChecker<'a, 'tcx> {
         self.super_projection_elem(local, proj_base, elem, context, location);
     }
 
-    fn visit_place(&mut self, place: &Place<'tcx>, _: PlaceContext, _: Location) {
+    fn visit_place(&mut self, place: &Place<'tcx>, cntxt: PlaceContext, _: Location) {
         // Set off any `bug!`s in the type computation code
         let _ = place.ty(&self.body.local_decls, self.tcx);
+
+        if self.mir_phase >= MirPhase::Derefered
+            && place.projection.len() > 1
+            && cntxt != PlaceContext::NonUse(VarDebugInfo)
+        {
+            assert!(!place.projection[1..].contains(&ProjectionElem::Deref), "{place:#?}");
```

Comment:
Is there any reason not to use `self.fail` here?

Reply:
It gave horrendous wall of text as error I thought assert's error looked cleaner and concise.

Reply:
Hmm, honestly, I'm not so concerned by the wall of text. This is an ICE after all, it doesn't really have to have good UX, just needs to have all the necessary information. Especially the part about the location being included in the ICE seems really useful.

Reply:
I changed it to `self.fail` after trying it out it is indeed more useful than basic assert